File: backend/services/perception_lab/families/registry.py
"""One explicit registry; fixed imports, no discovery and no model loading."""
from __future__ import annotations

from . import colour, depth, illumination, orientation_flow, surface_form, surface_pattern
from .base import FamilyModule

FAMILY_KEYS = ("colour", "illumination", "surface_pattern", "orientation_flow", "depth", "surface_form")
SLOTS = (colour.MODULE, illumination.MODULE, surface_pattern.MODULE,
         orientation_flow.MODULE, depth.MODULE, surface_form.MODULE)


def checked_slots(slots: tuple[FamilyModule, ...] = SLOTS) -> dict[str, FamilyModule]:
    if tuple(slot.family for slot in slots) != FAMILY_KEYS:
        raise ValueError("family slots must be the six fixed keys in registry order")
    seen_forms, seen_ops, seen_producers = set(), set(), set()
    for slot in slots:
        if not slot.label or not slot.reason or not isinstance(slot.available, bool):
            raise ValueError(f"malformed family declaration {slot.family}")
        if not slot.available and (slot.forms or slot.operations or slot.producers or slot.models):
            raise ValueError(f"unavailable family {slot.family} declares live producers or forms")
        forms = {form.key for form in slot.forms}
        producers = set(slot.producers or {})
        if len(forms) != len(slot.forms) or len(producers) != len(slot.producers or {}):
            raise ValueError(f"duplicate family-local declaration in {slot.family}")
        if slot.available and (not forms or not slot.operations or not producers):
            raise ValueError(f"available family {slot.family} has no complete instrument")
        for form in slot.forms:
            if not form.key.startswith(slot.family + ".") or not form.views or not form.quantity:
                raise ValueError(f"malformed form {form.key}")
            if form.key in seen_forms:
                raise ValueError(f"duplicate form {form.key}")
            seen_forms.add(form.key)
        for op in slot.operations:
            if (not op.key.startswith(slot.family + ".") or op.form_key not in forms
                    or op.producer_key not in producers or op.key in seen_ops
                    or any(not intent or len(intent) > 160 for intent in op.prompt_intents)):
                raise ValueError(f"malformed or duplicate operation {op.key}")
            if op.learned and not slot.models:
                raise ValueError(f"learned operation {op.key} has no admitted model declaration")
            seen_ops.add(op.key)
        for key, producer in (slot.producers or {}).items():
            if not key.startswith(slot.family + ".") or not callable(producer) or key in seen_producers:
                raise ValueError(f"malformed or duplicate producer {key}")
            seen_producers.add(key)
        for model in slot.models:
            if (not all(model.get(k) for k in ("key", "revision", "license", "checkpoint_digest"))
                    or type(model.get("admitted")) is not bool):
                raise ValueError(f"incomplete model declaration in {slot.family}")
        for dependency in slot.dependencies:
            if (not all(dependency.get(k) for k in ("package", "version", "license"))
                    or type(dependency.get("admitted")) is not bool):
                raise ValueError(f"incomplete dependency declaration in {slot.family}")
        if slot.available and any(not declaration["admitted"] for declaration in
                                  (*slot.models, *slot.dependencies)):
            raise ValueError(f"unadmitted model or dependency in {slot.family}")
    return {slot.family: slot for slot in slots}


REGISTRY = checked_slots()


def validate_field_artifact(artifact) -> None:
    family = artifact.identity.organ_family.value
    slot = REGISTRY.get(family)
    if slot is None or not slot.available:
        raise ValueError(f"{family} is not an available field family")
    payload = artifact.measurement.payload
    op = next((item for item in slot.operations if item.key == artifact.identity.operation), None)
    if op is None or payload is None or payload.form_key != op.form_key:
        raise ValueError("sample grid operation/form does not match its family slot")
    if payload.manifest["metadata"]["source_digest"] != artifact.provenance.source_image_digest:
        raise ValueError("sample grid source digest differs from artifact provenance")

File: frontend/src/perceptionLab/families/registry.js
import { SLOT as colour } from './colour';
import { SLOT as illumination } from './illumination';
import { SLOT as surfacePattern } from './surface_pattern';
import { SLOT as orientationFlow } from './orientation_flow';
import { SLOT as depth } from './depth';
import { SLOT as surfaceForm } from './surface_form';

export const FAMILY_KEYS = Object.freeze([
    'colour', 'illumination', 'surface_pattern', 'orientation_flow', 'depth', 'surface_form',
]);

export function checkedFamilySlots(slots = [colour, illumination, surfacePattern,
    orientationFlow, depth, surfaceForm]) {
    if (slots.length !== 6 || slots.some((slot, index) => slot.family !== FAMILY_KEYS[index])) {
        throw new Error('six fixed family slots must appear in registry order');
    }
    const forms = new Set();
    const operations = new Set();
    slots.forEach((slot) => {
        if (!slot.label || !slot.reason || typeof slot.Panel !== 'function'
            || typeof slot.available !== 'boolean' || !Array.isArray(slot.forms)
            || !Array.isArray(slot.operations) || !Array.isArray(slot.views)
            || !Array.isArray(slot.promptIntents)) {
            throw new Error(`malformed family slot ${slot.family}`);
        }
        if (!slot.available && (slot.forms.length || slot.operations.length)) {
            throw new Error(`unavailable family ${slot.family} declares live forms`);
        }
        if (slot.available && (!slot.forms.length || !slot.operations.length || !slot.views.length)) {
            throw new Error(`available family ${slot.family} has no complete panel contract`);
        }
        if (slot.promptIntents.some((intent) => typeof intent !== 'string'
            || !intent.trim() || intent.length > 160)) {
            throw new Error(`malformed bounded prompt intent in ${slot.family}`);
        }
        slot.forms.forEach((form) => {
            if (!form.key?.startsWith(`${slot.family}.`) || !form.label
                || !form.quantity || !Array.isArray(form.views) || !form.views.length
                || form.views.some((view) => !slot.views.includes(view)) || forms.has(form.key)) {
                throw new Error(`malformed or duplicate form ${form.key}`);
            }
            forms.add(form.key);
        });
        slot.operations.forEach((op) => {
            if (!op.key?.startsWith(`${slot.family}.`) || operations.has(op.key)
                || !forms.has(op.form_key) || !op.producer_key?.startsWith(`${slot.family}.`)
                || !op.label || !op.parameters || typeof op.parameters !== 'object') {
                throw new Error(`malformed or duplicate operation ${op.key}`);
            }
            operations.add(op.key);
        });
    });
    return Object.freeze(Object.fromEntries(slots.map((slot) => [slot.family, slot])));
}

export const FAMILY_SLOTS = checkedFamilySlots();
